[Upgade Assistant] Enable UA for minor upgrades

Bamieh opened the pull request:
# Summary

UA is always enabled, the customer needs to check if their cluster needs an upgrade before they start addressing warnings and errors. This way they can also proactively address errors and warnings preventing upgrades before a new stack release is even available.

Currently, we are not pinging an external API to check if the cluster is not at the latest stack version hence this step must be manually done by the users.

- [x] Enable UA for minor upgrades
- [x] New UA design changes
- [x] Update docs links
- [x] Connect ES upgrade readiness apis
- [x] Review new UA interface copy
- [x] Unit and functional tests

## Details and screenshots:

### Updated UA text to adapt to any release upgrade not just `8.latest`
The documentation links will redirect the user to the `current` version of the stack instead of `8.0`. This way the user will check if they are not on the latest version and follow through with the upgrade steps.

<img width="949" alt="image" src="https://user-images.githubusercontent.com/6191849/196455440-c464b95a-232d-4963-8eb1-047f7f77db09.png">

<img width="1266" alt="image" src="https://user-images.githubusercontent.com/6191849/199257940-2319bec6-f903-4ecc-b0ab-c1e0e6144fc3.png">


### Upgrade readiness checks
The ES upgrade readiness checks are integrated into their deprecations API.
Initially we had designs to keep this in a separate step, however merging the checks into the already existing third step provides a better UX and a shorter barrier to upgrade

New checks:
- Critical deprecation warning for potentially exceeding maximum open shards setting https://github.com/elastic/elasticsearch/issues/85702 
- Critical deprecation warning for low disk watermark https://github.com/elastic/elasticsearch/pull/83544

### Migrating system indices
Migrating system indices are only required during major upgrades. This step will be marked as complete to the users unless they are upgrading to a major version. On non-major version upgrades, we will hide this step completely.

# Manual testing

Since the ES changes are using the same deprecation API there is no need to add tests from our side to check that they are getting triggered correctly outside a couple of surface-level checks.

## Minor version upgrade

1. Launch Kibana
2. Go to advanced settings / upgrade assistant
3. Check upgrade steps

## Upgrade Readiness
The ES Upgrade readiness checks are integrated into the ES deprecations API and will show up on the Elasticsearch deprecations section inside the Upgrade Assistance.

## Low disk watermark

1. Reduce the low watermark disk threshold to mimic disk usage exceeding low watermark. Note that for ES running in dev mode the disk threshold is disabled hence we'd need to enable it as well.

```
PUT _cluster/settings
{
  "transient": {
    "cluster.routing.allocation.disk.watermark.low" : "30%",
    "cluster.routing.allocation.disk.threshold_enabled": true,
    "cluster.info.update.interval": "10s"
  }
}
```

2. Call the ES deprecations API
```
GET /_migration/deprecations
```

3. Check the result
```
{
  "cluster_settings": [],
  "node_settings": [
    {
      "level": "critical",
      "message": "Disk usage exceeds low watermark",
      "url": "https://ela.st/es-deprecation-7-disk-watermark-exceeded",
      "details": "Disk usage exceeds low watermark, which will prevent reindexing indices during upgrade. Get disk usage on all nodes below the value specified in cluster.routing.allocation.disk.watermark.low (nodes impacted: [Bamieh-MBP])",
      "resolve_during_rolling_upgrade": false
    }
  ],
  "index_settings": {},
  "ml_settings": []
}
```

4. Open the upgrade assistant to check the deprecation

<img width="664" alt="image" src="https://user-images.githubusercontent.com/6191849/199257074-e3073413-6825-4e57-b37b-29a8a5382003.png">

<img width="1725" alt="image" src="https://user-images.githubusercontent.com/6191849/199257025-53223102-a8d6-4934-a6d9-7f42e8df4e9c.png">

## Total shards check

1. Reduce the max shards per node to mimic the deprecation when the cluster has too many shards
```
PUT /_cluster/settings
{
  "persistent" : {
    "cluster.max_shards_per_node" : "9"
  }
}
```

2. Call the ES deprecations API
```
GET /_migration/deprecations
```

3. Check the result
```
{
  "cluster_settings": [
    {
      "level": "warning",
      "message": "The cluster has too many shards to be able to upgrade",
      "url": "https://ela.st/es-deprecation-8-shard-limit",
      "details": "Upgrading requires adding a small number of new shards. There is not enough room for 10 more shards. Increase the cluster.max_shards_per_node setting, or remove indices to clear up resources.",
      "resolve_during_rolling_upgrade": false
    }
  ],
  "node_settings": [],
  "index_settings": {},
  "ml_settings": []
}
```

4. Open the upgrade assistant to check the deprecation
<img width="662" alt="image" src="https://user-images.githubusercontent.com/6191849/199254045-c49a52f5-6021-4020-9131-51c72ed853e5.png">

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/6191849/199259679-2f06287b-6bd0-4f15-814e-4d509deda0f2.png">

florent-leborgne commented:
@Bamieh ran me through the copy changes and it looks good to me. cc @gchaps

thomheymann reviewed (approved):
security changes LGTM

KOTungseth reviewed (approved):
On behalf of @florent-leborgne, I approve this PR.

alisonelizabeth reviewed (commented):
Thanks so much for working on this @Bamieh! I'm excited to see UA getting some attention again 😄 I left a couple minor comments in the code.

I also had a general question about turning UA back on. One of the reasons we disabled it post-8.0 was because the reindex action includes some logic that is [specific to the 8.0](https://github.com/elastic/kibana/blob/main/x-pack/plugins/upgrade_assistant/server/lib/reindexing/index_settings.ts) upgrade, and we were concerned a user could get into a bad situation if we didn't account for something. We ran into a couple bugs in 7.x (pre-8.0 UA) where something similar happened (https://github.com/elastic/kibana/issues/40564, https://github.com/elastic/kibana/issues/57656). Is this still a valid concern?

There was also some logic added around [handling old ML job model snapshots](https://github.com/elastic/kibana/blob/main/x-pack/plugins/upgrade_assistant/server/routes/ml_snapshots.ts) that I'm not sure is still relevant or not for minor upgrades. @droberts195 maybe you can provide some feedback here?

droberts195 commented:
> There was also some logic added around [handling old ML job model snapshots](https://github.com/elastic/kibana/blob/main/x-pack/plugins/upgrade_assistant/server/routes/ml_snapshots.ts) that I'm not sure is still relevant or not for minor upgrades. @droberts195 maybe you can provide some feedback here?

It's not needed for minor version upgrades.

It would be nice not to completely delete all the code, as ML model snapshot upgrade is bound to be needed again at some point in the future.  But please hide it from the visible UI for the time being.

droberts195 commented on x-pack/plugins/upgrade_assistant/server/routes/ml_snapshots.ts:
```diff
@@ -330,8 +331,25 @@ export function registerMlSnapshotRoutes({
-    versionCheckHandlerWrapper(async ({ core }, request, response) => {
+    versionCheckHandlerWrapper(async ({ core, ...rest }, request, response) => {
       try {
+        /**
+         * Always return false if featureSet.mlSnapshots is set to false
+         * This disables showing the ML deprecations in the UA UI
+         * 
+         * This config should be set to false only on the `x.last` versions
+         * ML `upgrade_mode` can be enabeld from outside Kibana, the prurpose
+         * of this feature guard is to hide all ML related deprecations from the end user 
+         * until the next major upgrade.
+         */
```
```suggestion
        /**
         * Always return false if featureSet.mlSnapshots is set to false
         * This disables possibly showing a needless warning about ML
         * upgrade mode when there's never a need to upgrade ML job
         * snapshots in minor version upgrades.
         * 
         * This config should be set to false only on the `x.last` versions, or when
         * the constant `MachineLearningField.MIN_CHECKED_SUPPORTED_SNAPSHOT_VERSION`
         * is incremented to something higher than 7.0.0 in the Elasticsearch code.
         */
```

I made some suggestions to edit the comment.

Setting `mlUpgradeModeEnabled` to `false` will stop https://github.com/elastic/kibana/blob/47ebe8bef760828567aaa167c7e20fc846c10a9d/x-pack/plugins/upgrade_assistant/public/application/components/es_deprecations/deprecation_types/ml_snapshots/flyout.tsx#L215 from being shown, which is good since that message is irrelevant until we actually want to start doing ML snapshot upgrades again.

However, `mlUpgradeModeEnabled` set to `false` _will_ allow actual ML snapshot deprecation messages to be shown - see https://github.com/elastic/kibana/blob/47ebe8bef760828567aaa167c7e20fc846c10a9d/x-pack/plugins/upgrade_assistant/public/application/components/es_deprecations/deprecation_types/ml_snapshots/flyout.tsx#L246. So what stops that happening? The answer is https://github.com/elastic/elasticsearch/blob/34ae3dd5b13e01878e07ecd494d0be2c92676a8f/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MachineLearningField.java#L50.

So I think everything should work with this code change as it is, but it's just that the explanation for _why_ it will work needs clarifying.

When we want to enable ML model snapshot deprecation warnings again we need to change https://github.com/elastic/elasticsearch/blob/34ae3dd5b13e01878e07ecd494d0be2c92676a8f/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MachineLearningField.java#L50 and also `featureSet.mlSnapshots` in the same release.

droberts195 replied:
> Do you think it makes sense to add this logic on top as well?

Yes, it's probably safest to do this.

droberts195 reviewed (approved):
ML changes LGTM

(I didn't look at anything outside of the ML changes)

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/README.md:
```diff
@@ -2,7 +2,7 @@
 ## About
 
-Upgrade Assistant helps users prepare their Stack for being upgraded to the next major. It will only be enabled on the last minor before the next major release. This is controlled via the config: `xpack.upgrade_assistant.readonly` ([#101296](https://github.com/elastic/kibana/pull/101296)).
+Upgrade Assistant helps users prepare their Stack for being upgraded to the next version of the Elastic stack.
```
Do you think you can update the readme with information around the `featureSet` config?

alisonelizabeth commented on x-pack/plugins/upgrade_assistant/server/config.ts:
```diff
@@ -4,98 +4,53 @@
 // -------------------------------
-// >= 8.x
+// >= 8.6 UA is always enabled to guide stack upgrades
+// even for minors releases.
```
```suggestion
// even for minor releases.
```

alisonelizabeth reviewed (approved):
Latest changes LGTM. Thanks for addressing my feedback!

I'm going to go ahead and approve to unblock you, but I have two more requests before merging 😄 
1 - Do you mind updating the readme with some information about the new `featureSet` config?
2 - I just remembered we also have some API integration tests for UA. I think most of them only run on 7.x ([example](https://github.com/elastic/kibana/blob/main/x-pack/test/api_integration/apis/upgrade_assistant/es_deprecations.ts#L16)). Can you take a [quick look through them](https://github.com/elastic/kibana/tree/main/x-pack/test/api_integration/apis/upgrade_assistant) and see if it makes sense to enable any of them now? Feel free to reach out if you have any questions. Thanks!

alisonelizabeth commented:
@Bamieh in addition to the API integration tests, I see we also have some [functional tests ](https://github.com/elastic/kibana/tree/main/x-pack/test/functional/apps/upgrade_assistant)that are being skipped right now. Can you take a look at those as well? Thanks!

Bamieh commented:
@alisonelizabeth Thanks :elasticheart:

- [x] I've enabled the api_integration tests that make sense (13 suites enabled)
- [x] I've updated the documentation and added a section for the `featureSet` configs
- [x] I've enabled all the functional tests and added additional tests for cluster upgrade readiness
